Remove the `pg_sql_graph_magic!()` macro.

eeeebbbbrrrr opened the pull request:
The `pg_sql_graph_magic!()` did some shenanigans related to loading the extension ".control" file at compile-time and embedded it in the compiled artifact.  This was convoluted and actually unnecessary as the location where that information was used (in `schema.rs`) can easily access the ".control" file at runtime.

This PR removes that macro and its usages.

The `fn __pgrx__marker()` function is still required, however.  It's necessary to ensure that the `rustc` build we do for the `pgrx_embed` binary for schema generation properly links so that the other (dynamically referenced) `__pgrx_internals_XXX` symbols can be found.  The function, however, is now a noop.

This is related to #1546, but is not a solution.  It should just make solving that request easier.

eeeebbbbrrrr committed:
Remove the `pg_sql_graph_magic!()` macro.

The `pg_sql_graph_magic!()` did some shenanigans related to loading the extension ".control" file at compile-time and embedded it in the compiled artifact.  This was convoluted and actually unnecessary as the location where that information was used (in `schema.rs`) can easily
access the ".control" file at runtime.

This PR removes that macro and its usages.

The `fn __pgrx__marker()` function is still required, however.  It's necessary to ensure that the `rustc` build we do for the `pgrx_embed` binary for schema generation properly links so that the other (dynamically referenced) `__pgrx_internals_XXX` symbols can be found.  The function, however, is now a noop.

eeeebbbbrrrr commented on pgrx-tests/src/framework.rs:
```diff
@@ -312,7 +312,7 @@ fn install_extension() -> eyre::Result<()> {
         .arg(pg_config.path().ok_or(eyre!("No pg_config found"))?)
         .stdout(Stdio::inherit())
-        .stderr(Stdio::piped())
+        .stderr(Stdio::inherit())
```
maybe out of place in this PR, but I burned well over an hour trying to figure out why the `cargo-pgrx` output was missing, leading me to believe schema generation wasn't even happening during `cargo test`.

workingjubilee replied:
Any diagnostics changes necessary to uhhh write the PR are appropriate, I think, barring like a complete rewrite of framework.rs

eeeebbbbrrrr committed:
woah.  need to get that envvar at runtime, not compile time!

eeeebbbbrrrr commented on pgrx/src/lib.rs:
```diff
@@ -182,7 +182,13 @@ const _: () = {
+        // A marker function which must exist in the root of the extension for proper linking by the
+        // "pgrx_embed" binary during `cargo-pgrx schema` generation.
+        #[doc(hidden)]
+        pub fn __pgrx_marker() {
```
this should probably be `#[inline(never)`].

I haven't been able to cause it to happen, but it seems to me that the compiler's DCE passes could somehow cause the `pgrx_embed` binary to not link to the .so if the function and its call point in `pgrx_embed::main()` were removed.

EDIT:  It does **NOT** also need to be `#[no_mangle]` as it's referenced directly, not indirectly through a `extern "Rust" {}` declaration like all the other `__pgrx_internals_XXX` functions are.

workingjubilee replied:
That is very astute, as Saethlin has been working extensively on making the compiler do _very aggressive_ inlining, because idiomatic Rust uses lots of tiny functions.

eeeebbbbrrrr committed:
code review cleanup

eeeebbbbrrrr committed:
code review cleanup

eeeebbbbrrrr committed:
minor refactor to move .control file variable substitutions to `control_file.rs`

eeeebbbbrrrr committed:
come on, `cargo fmt`.  seriously

eeeebbbbrrrr commented on pgrx-sql-entity-graph/src/control_file.rs:
```diff
@@ -83,13 +96,13 @@ impl ControlFile {
                 .get("relocatable")
                 .ok_or(ControlFileError::MissingField { field: "relocatable" })?
-                == &"true",
+                == "true",
```
`cargo clippy` got upset once I touched this file.  Same below on lines 103 and 105.

eeeebbbbrrrr committed:
carry cargo's `CARGO_PKG_VERSION` value through to the `pgrx_embed` binary as our own

workingjubilee commented:
For the audience's benefit: This cannot be landed as-is because it breaks `cargo pgrx schema`, which is somewhat necessary for diagnostic purposes.

eeeebbbbrrrr committed:
in order to propertly support the `@CARGO_VERSION@` variable that can be in the ".control" file, we need to pluck the version number out of the manifest and pass it through to the `pgrx_embed` binary as an envvar.

Also add a CI test for running `cargo pgrx schema`, which is how I manually discovered this problem.

eeeebbbbrrrr commented:
> For the audience's benefit: This cannot be landed as-is because it breaks `cargo pgrx schema`, which is somewhat necessary for diagnostic purposes.

ee63bf9 fixes that.  ;)

workingjubilee commented on cargo-pgrx/src/command/schema.rs:
```diff
@@ -545,6 +568,12 @@ fn compute_sql(profile: &CargoProfile, package_name: &str) -> eyre::Result<()> {
     command.stderr(Stdio::inherit());
 
+    // pass the package version through as an environment variable
+    command.env(
+        "PGRX_PKG_VERSION",
+        manifest.package_version().expect("`Cargo.toml` is missing the package version property"),
+    );
```
I would prefer if this checked if `std::env::var("CARGO_PKG_VERSION")` is set and only when it is unset do the needful, and when we do that, **lie**: set Cargo's env vars for it, and only set the env vars we need. This avoids adding another quasi-stable quasi-unstable undocumented interface to our code.

eeeebbbbrrrr replied:
okay.

eeeebbbbrrrr committed:
more cleanups around `CARGO_PKG_VERSION`

eeeebbbbrrrr committed:
seriously clippy.

eeeebbbbrrrr committed:
CI:  UGH

eeeebbbbrrrr committed:
I swear.  Something about CI causes my brain to shut off, causing endless frustration.

eeeebbbbrrrr commented:
It’s green!

workingjubilee commented:
Holy shit!

workingjubilee commented on pgrx-sql-entity-graph/src/control_file.rs:
```diff
@@ -43,18 +45,46 @@ pub struct ControlFile {
+        fn do_var_replacements(mut input: String) -> Result<String, ControlFileError> {
+            const CARGO_VERSION: &str = "@CARGO_VERSION@";
+
+            // endeavor to not require external values if they're not used by the input
+            if input.contains(CARGO_VERSION) {
+                input = input.replace(
+                    CARGO_VERSION,
+                    &std::env::var("CARGO_PKG_VERSION").map_err(|_| {
+                        ControlFileError::MissingEnvvar("CARGO_PKG_VERSION".to_string())
+                    })?,
+                );
+            }
```
This does introduce a slight behavior divergence, but that seems fine, since the main validation effort is done via cargo-pgrx.

workingjubilee commented on cargo-pgrx/src/command/schema.rs:
```diff
@@ -391,9 +399,20 @@ fn compute_codegen(
+            .to_str()
+            .expect(".control file filename should be valid UTF8");
         let mut out = quote::quote! {
+            // call the marker.  Primarily this ensures that rustc will actually link to the library
+            // during the "pgrx_embed" build initiated by `cargo-pgrx schema` generation
+            #lib_name_ident::__pgrx_marker();
```
Perhaps we should use `pub extern "C" fn` on this marker? rustc is more reluctant to play around with obviously-for-external-consumption interfaces. It's not something we need to fix now, however, just musing aloud.

workingjubilee commented on cargo-pgrx/src/command/schema.rs:
```diff
@@ -391,9 +399,20 @@ fn compute_codegen(
+
             let mut entities = Vec::new();
-            let control_file_entity = ::pgrx::pgrx_sql_entity_graph::SqlGraphEntity::ExtensionRoot(::#lib_name_ident::__pgrx_marker());
+            let control_file_path = std::path::PathBuf::from(#control_file_path);
```
I wonder if there's some cursed set of cases that results in this being a malformed UTF-8 string? ...but if someone seriously asks for supporting non-UTF-8 paths in our greenfield extension build system, I'm probably going to tell them exactly where they can shove that request. Maybe we should be simplifying our code by using [camino](https://docs.rs/camino/latest/camino/).

workingjubilee commented on pgrx-sql-entity-graph/src/control_file.rs:
```diff
@@ -43,18 +45,46 @@ pub struct ControlFile {
     /// use pgrx_sql_entity_graph::ControlFile;
     /// # fn main() -> eyre::Result<()> {
-    /// let context = include_str!("../../pgrx-examples/custom_types/custom_types.control");
+    /// # // arrays.control chosen because it does **NOT** use the @CARGO_VERSION@ variable
+    /// let context = include_str!("../../pgrx-examples/arrays/arrays.control");
```
Yeah this seems fair. We have other tests to guarantee this works.

workingjubilee commented on pgrx/src/lib.rs:
```diff
@@ -259,39 +266,6 @@ macro_rules! pg_magic_func {
-/// This macro should only be directly called in advanced use cases.
-///
-/// </pre></div>
-#[macro_export]
-macro_rules! pg_sql_graph_magic {
-    () => {
-        // A marker which must exist in the root of the extension.
-        #[doc(hidden)]
-        pub fn __pgrx_marker() -> $crate::pgrx_sql_entity_graph::ControlFile {
-            use ::core::convert::TryFrom;
-            let package_version = env!("CARGO_PKG_VERSION");
-            let context = include_str!(concat!(
-                env!("CARGO_MANIFEST_DIR"),
-                "/",
-                env!("CARGO_CRATE_NAME"),
-                ".control"
-            ))
-            .replace("@CARGO_VERSION@", package_version);
-
-            let control_file =
-                $crate::pgrx_sql_entity_graph::ControlFile::try_from(context.as_str())
-                    .expect("Could not parse control file, is it valid?");
-            control_file
-        }
-    };
-}
```
Excellent.

workingjubilee reviewed (approved):
Just some comments for future reference if I ever have to come back to this PR. It looks good! Moving build complexity out of the library is great.
